feat(ex_cmds): support ranged :lua

luukvbaal opened the pull request:
:{range}lua executes the specified lines in the current buffer as
Lua code, regardless of its extension or 'filetype'.

Close #27103

zeertzjq reviewed (commented):
`{chunk}` being ignored when `[range]` is present may be a bit strange. Give an error in that case?

zeertzjq commented on src/nvim/lua/executor.c:
```diff
@@ -1649,6 +1649,17 @@ bool nlua_is_deferred_safe(void)
+    }
+    return;
+  } else if (*eap->arg == NUL) {
+    emsg(_(e_invarg));
```
```suggestion
    emsg(_(e_argreq));
```

noahfrederick commented on runtime/doc/lua.txt:
```diff
@@ -242,10 +242,12 @@ The Lua print() function redirects its output to the Nvim message area, with
-    Executes Lua chunk {chunk}. If {chunk} starts with "=" the rest of the
-    chunk is evaluated as an expression and printed. `:lua =expr` or `:=expr` is
-    equivalent to `:lua print(vim.inspect(expr))`.
+:[range]lua {chunk}
```
If `[range]` and `{chunk}` are mutually exclusive, I would expect `:[range]lua` and `:lua {chunk}` to each be documented in separate paragraphs (c.f., `:help :read`).

luukvbaal committed:
feat(ex_cmds): support ranged :lua

:{range}lua executes the specified lines in the current buffer as
Lua code, regardless of its extension or 'filetype'.

noahfrederick commented on runtime/doc/lua.txt:
```diff
@@ -253,6 +253,10 @@ arguments separated by " " (space) instead of "\t" (tab).
 <    To see the LuaJIT version: >vim
         :lua =jit.version
+
+:[range]lua
```
```suggestion
:{range}lua
```

Since the range is required for this usage.

luukvbaal commented on runtime/doc/lua.txt:
```diff
@@ -253,6 +253,10 @@ arguments separated by " " (space) instead of "\t" (tab).
         :lua =jit.version
+
+:[range]lua
+    Executes the |[range]| in the current buffer aso Lua code, regardless of
+    its extension or the current |'filetype'|.
```
I think it's worth keeping the `|[range]|` link at least.

clason replied:
Ah, no, that's not the only reason. You might want to evaluate a snippet in a Lua file as well (e.g., for simple debugging of a condition in a script).

luukvbaal replied:
But you can already do that with `:[range]source`.

clason replied:
Fair; in this case this is indeed useful information -- but it might be more useful to make explicit the comparison to `:source` (e.g., "unlike `:[range]source`, this evaluates the selected lines as Lua independent of `'filetype'`")?

luukvbaal replied:
Yeah that's what I was thinking as well.

rudiejd commented:
should we update the "lua interface" section of `:h vim-differences` here? https://github.com/neovim/neovim/blob/4e59422e1d4950a3042bad41a7b81c8db4f8b648/runtime/doc/vim_diff.txt#L494

justinmk commented on src/nvim/lua/executor.c:
```diff
@@ -1649,6 +1649,15 @@ bool nlua_is_deferred_safe(void)
+    if (eap->addr_count > 0 && *eap->arg == NUL) {
+      cmd_source_buffer(eap, true);
+    } else {
+      semsg(_(e_invarg2), "one of {chunk} and [count] is required");
```
Let's use the same phrasing as https://github.com/neovim/neovim/blob/4e59422e1d4950a3042bad41a7b81c8db4f8b648/src/nvim/eval/typval.h#L441

```suggestion
      semsg(_(e_invarg2), "{chunk} or [count] is required");
```

zeertzjq replied:
I don't think this is a good idea, because both `{chunk}` and `{range}` can be specified, which needs to be rejected.

luukvbaal replied:
Go with the previous phrasing then? Or just

    either {chunk} or {range} is required

luukvbaal replied:
Or do we need separate errors:

    {chunk} or {range} required

and

    only one of {chunk} and {range} allowed

zeertzjq replied:
`exactly one of`?

luukvbaal replied:
Works for me. Is there no precedent for this kind of message?

zeertzjq replied:
Hmm, `nvim_create_autocmd` has two separate error messages for mutually exclusive `command` and `callback`.

justinmk reviewed (approved):
Great to see this!

zeertzjq commented on runtime/doc/lua.txt:
```diff
@@ -253,6 +253,11 @@ arguments separated by " " (space) instead of "\t" (tab).
+    Executes the |[range]| in the current buffer as Lua code. Unlike |:source|,
+    this will execute the specified lines regardless of the extension or
+    |'filetype'| of the buffer.
 <
```
```suggestion

```

zeertzjq commented on runtime/doc/lua.txt:
```diff
@@ -253,6 +253,11 @@ arguments separated by " " (space) instead of "\t" (tab).
+    Executes the |[range]| in the current buffer as Lua code. Unlike |:source|,
+    this will execute the specified lines regardless of the extension or
+    |'filetype'| of the buffer.
 <
```
Remove this `<`

luukvbaal replied:
Oh oops, not remove but place this block below it right?

zeertzjq replied:
Either works.

zeertzjq commented on runtime/doc/lua.txt:
```diff
@@ -254,6 +254,11 @@ arguments separated by " " (space) instead of "\t" (tab).
 <    To see the LuaJIT version: >vim
         :lua =jit.version
 <
+
```
Move this blank line below the block

zeertzjq commented on src/nvim/lua/executor.c:
```diff
@@ -1649,6 +1649,15 @@ bool nlua_is_deferred_safe(void)
+    if (eap->addr_count > 0 && *eap->arg == NUL) {
+      cmd_source_buffer(eap, true);
+    } else {
+      semsg(_(e_invarg2), "{chunk} or [count] is required");
```
```suggestion
      semsg(_(e_invarg2), "{chunk} or {range} is required");
```

zeertzjq commented:
Related #7177

justinmk commented:
> This sort of conflicts with #7177 as having both will be quite confusing.

Good point. This PR is more important though: piping _buffer text_ to `:lua <code>` isn't very convenient because `vim.firstline` and `vim.lastline` are verbose to type out, thus it's uncommon. Whereas executing a Lua snippet is very common.

If there is a strong use-case for `:{range}lua foo(vim.firstline)` we can always enable that behavior when both are present (though it's a bit clever/confusing as you noted).
